pkg/controller/job/job_controller.go

```diff
@@ -448,6 +448,16 @@ func (jm *Controller) deletePod(logger klog.Logger, obj interface{}, final bool)
 	jm.enqueueSyncJobBatched(logger, job)
 }
 
+func (jm *Controller) addJob(logger klog.Logger, obj interface{}) {
+	jm.enqueueSyncJobImmediately(logger, obj)
+	jobObj, ok := obj.(*batch.Job)
+	if ok {
+		if controllerName := managedByExternalController(jobObj); controllerName != nil {
+			metrics.JobByExternalControllerTotal.WithLabelValues(*controllerName).Inc()
+		}
+	}
+}
+
 func (jm *Controller) updateJob(logger klog.Logger, old, cur interface{}) {
```

Comment:
> Can you please review the updateJob, deleteJob and the pod event handlers?
> It seems to me that a job controller should also not react to the pods of unmanaged jobs. In particular to the batch.kubernetes.io/job-tracking finalizer. These should be resolved by the external controller.
> updateJob also reacts to ActiveDeadlineSeconds and triggers the queue which would be better to skip sooner

Good point, let's review them one by one.

Regarding finalizers (which are accounted in a couple of functions here) they are reserved to be used by the core k8s Job controller. I expect the other controllers not to use them, and potentially use finalizers with different names. 

Even if they are in use deleting the finalizers from a finished job is ok I believe: https://github.com/kubernetes/kubernetes/blob/e457683e6cb740656f389932fb564a6ab45be759/pkg/controller/job/job_controller.go#L472-L474.  cc @alculquicondor

1. **addJob**

This one only requeues (skipped).


2. **updateJob**

Aside from the finalizers cleanup we check the `ActiveDeadlineSeconds` which is unnecessary here. This is another path we can currently step into `syncJob` even if `managed-by` is used. I missed that originally. 

I think skipping checking of `ActiveDeadlineSeconds` as a performance optimization make sense, so that we don't put items into the queue (eating memory). It might be helpful when many jobs use this field.

3. **deleteJob**

This one only requeues (skipped) and calls `cleanupPodFinalizers` which should be desired to cleanup the reserved finalizers for orphaned pods.

4. **addPod**

- calling `recordFinishedPodWithTrackingFinalizer` I assume is fine, this is built-in finalizer anyway
- calling `jm.expectations.CreationObserved(logger, jobKey)` is no op if the expectations are not setup (they are only setup in syncJob)
- calling `jm.enqueueSyncJobBatched(logger, job)` is skipped inside
- calling `if hasJobTrackingFinalizer(pod) { jm.enqueueOrphanPod(pod) }` I assume is fine to cleanup orphan pods
- calling `for _, job := range jm.getPodJobs(pod) { jm.enqueueSyncJobBatched(logger, job) }` is skipped inside 

5. **updatePod** 

- calling `recordFinishedPodWithTrackingFinalizer(oldPod, curPod)` is just recording the metric, which we do regardless of the controller
- calling `finalizerRemovalObserved` is no op if the expectations are not initialized
- calling `jm.enqueueSyncJobBatched(logger, job)` is skipped inside
- calling `if hasJobTrackingFinalizer(pod) { jm.enqueueOrphanPod(pod) }` I assume is fine to cleanup orphan pods

6. **deletePod** 
- calling `recordFinishedPodWithTrackingFinalizer(pod, nil)` is just recording the metric, indepdendently of the controller
- calling `if hasJobTrackingFinalizer(pod) { jm.enqueueOrphanPod(pod) }` I assume is fine to cleanup orphan pods
- calling `jm.expectations.DeletionObserved(logger, jobKey)` is no-op as the expectations aren't set
- calling `jm.finalizerExpectations.finalizerRemovalObserved(logger, jobKey, string(pod.UID))` same
- calling `jm.enqueueSyncJobBatched(logger, job)` is skipped inside

So overall, unless I'm missing something it is all safe. One place possibly calling for optimization is the check for `ActiveDeadlineSeconds`. Let me know what you think. cc @alculquicondor ? I will apply the suggestion later today if no other voices.

Reply:
Actually, I just realized we have another use of `jm.queue.AddAfter` related to `ActiveDeadlineSeconds`. I would suggest to just replace `jm.queue.AddAfter` with `enqueueSyncJobInternal`. Getting the skip internally then. I will push it in this PR, but we may extract it also as a preparatory refactoring.

Reply:
After the second look this requeue only happens when `spec.ActiveDeadlineSeconds` is updated by the user. I think it is ok to reque in this case as it should be rare event. The sync will be skipped inside `syncJob`. So, I think it is better not to complicate the code for very tiny performance optimization in rare case. Let me know what you think.

Reply:
@mimowo mostly good assesment, but have a couple of remarks

updateJob:
deleteJob:
- we should also consider skipping `cleanupPodFinalizers` as it calls `enqueueOrphanPod`


addPod:
updatePod:
deletePod:
- `recordFinishedPodWithTrackingFinalizer` would record the `terminated_pods_tracking_finalizer_total` metric multiple times when multiple controllers are running
- we should also consider skipping `if hasJobTrackingFinalizer(pod) { jm.enqueueOrphanPod(pod) }`
- `for _, job := range jm.getPodJobs(pod) { jm.enqueueSyncJobBatched(logger, job) }` this could also loop the info logging

I agree that these are noop, but the orphan pod logic can send extra requests which is a bit problematic. We should limit the number of requests as much as we can and also with a larger number of controllers conflicts can happen.

In summary, there is quite a lot of logic in the event handlers and it is quite hard to reason about it. 

Can't we just simplify it and skip the jobs and pods as soon as we can in the flow? Even if we get it correct this time, any new addition to these handlers in the future could be difficult to get right.

I would expect any job controller to implement similar features as well, or possibly decide not to support the tracking finalizers for some strange reason. So we should be not miss any features by implementing the full skip.

Reply:
+1 for keeping the `ActiveDeadlineSeconds` / `AddAfter` logic simpler if it is an edge case.

Reply:
> we should also consider skipping cleanupPodFinalizers as it calls enqueueOrphanPod

The `batch.kubernetes.io/job-tracking` finalizers are in the `kubernetes.io` domain for a reason :) . Other external controllers should not use them, if they want some finalizes, use another controller-specific name. Similarly, nothing prevents currently users to add the finalizers to some pods not related to jobs. In that case the job controller would handle them as orphans. Also, since these pods are orphans thus we cannot reliably tell if they originated from jobs with managed-by or without (or any other CRD which used the reserved finalizers). I think cleaning them up, as we do, is ok.

 > recordFinishedPodWithTrackingFinalizer would record the terminated_pods_tracking_finalizer_total metric multiple times when multiple controllers are running
 
We will have exactly one built-in Job controller. Other external controllers should not bump internal kube-controller-manager metrics (not sure if they can technically).

> we should also consider skipping if hasJobTrackingFinalizer(pod) { jm.enqueueOrphanPod(pod) }

I think we should clean up the orphan pods as we currently do. Orphan means we don't have a reference to the owner, so we cannot tell if they originated from a job with managed-by or without.

> for _, job := range jm.getPodJobs(pod) { jm.enqueueSyncJobBatched(logger, job) } this could also loop the info logging

Yes, but as discussed in the other thread `enqueueSyncJobBatched` is already logging with the default level. Also, this situation is exceptional (just for orphan pods).

> I would expect any job controller to implement similar features as well, or possibly decide not to support the tracking finalizers for some strange reason. So we should not miss any features by implementing the full skip.

TBH I would rather express the sentiment that we are unlikely to see any controllers like this. The Job API is not that generic to fit needs of another controller. MultiKueue is a little bit special, because it mirrors 1:1 the status of the Job from a worker cluster, which is driven by the built-in controller.

Reply:
Yes it is a generic finalizer, but its implementation is in the job controller and not on admission. And we are turning off most of the logic for it - both assignment and removal (e.g. https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/job_controller.go#L1151). That is why I am not seeing useful to have the feature partially turned on.

> I think we should clean up the orphan pods as we currently do. Orphan means we don't have a reference to the owner, so we cannot tell if they originated from a job with managed-by or without.

I have missed this bit before. Nevertheless, we would still clean these up because we would not obtain the reference. The important part is that it would be nicely readable that we do not care about externally managed jobs or pods of these jobs.

So if we have the early skip. The resulting execution logic for pod handler would stay the same. Only change would be the skipped cleanup in the job handlers.

> We will have exactly one built-in Job controller. Other external controllers should not bump internal kube-controller-manager metrics (not sure if they can technically).

It can happen if you run a job controller twice. For example if you fork it.


I think it would be better to have a cleaner separation of concern. Even if it is okay for the current use case, it is hard to predict what all the future use cases will be and how many controllers will run at the same time.

@alculquicondor @soltysh can you please also give your opinion on this?

Reply:
Regarding the logic for handling orphan pods:
- the orphan pods are handled in code where we don't have the handle to the job to verify if it has managed-by or not (except when the job is being deleted or marked finished).  
- currently a user can create pods not owned by jobs (orphaned) and use the job tracking finalizer. The built-in job controller will clean up the pods, even though they don't originate from the built-in job controller.
- yes, it can be considered that cleanup of orphan pods with the finalizer is like the feature enabled partially, but it  is the part of the feature we want to keep enabled. The built-in controller takes care of cleaning orphan pods with the reserved finalizer, regardless of the origin of the pods.

Regarding the metrics by forked controllers:

The external controller if forked would not run inside `kube-controller-manager` so would not update built-in metrics (just its own metrics that happen to be named the same).

Reply:
> @alculquicondor @soltysh can you please also give your opinion on this?

We are giving full control over to external controller, including tracking full pod tracking, so we should NOT be doing anything with those pods. No extra calls - nothing, which would relate to managing any aspect of the job. Not even metrics, except for that managed by external controller one.

Reply:
> The external controller, if a fork, would not run inside `kube-controller-manager` so would not update built-in metrics (just its own metrics that happen to be named the same).

In a similar fashion, external controller should handle pod finalization, through its own dedicated finalizer. The built-in controller ignores the job completely.

Reply:
> We are giving full control over to external controller, including tracking full pod tracking, so we should NOT be doing anything with those pods. 

We are still discussing this on slack :) Indeed I could skip cleaning orphan pods with `batch.kubernetes.io/job-tracking` finalizer here: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/job_controller.go#L472-L474 (and in some other places where I have reference to the parent). However, there are places from which the orphan pods are cleaned and there is no reference to the parent, like here: https://github.com/kubernetes/kubernetes/blob/89cbd94e68ae80414dea7c14c1694e5e378a188d/pkg/controller/job/job_controller.go#L307-L309.

So far we did not distinguish these two cases so I'm hesitant to skip in a subset of cases.

Reply:
> In a similar fashion, external controller should handle pod finalization, through its own dedicated finalizer. The built-in controller ignores the job completely.

The only metrics updated from within even handlers is `terminated_pods_tracking_finalizer_total`. This metric is updated without lookup to parent, just based on the presence of the built-in finalizer, for example, from this place: https://github.com/kubernetes/kubernetes/blob/89cbd94e68ae80414dea7c14c1694e5e378a188d/pkg/controller/job/job_controller.go#L282. I don't have a guarantee to have a reference to the parent in this case. So far we updated the metric regardless if we had or not the reference to parent. Similarly, I'm concerned with distinguising three cases here: (1) don't have parent,  (2) have parent without managed by, (3) have parent with managed-by. I feel maintaining the behavior that the metric is bumped regardless of the reference to parent makes sense. It is only bumped for pods with built-in finalizer.

Reply:
@mimowo  We could try to resolve the controller reference the first thing we do in the pod handlers and see if it is our job. Even before the metrics.

The rest would flow as before. So even the orphaned pods without a controller would work as before.

Reply:
IIUC there is no discussion in this case:
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/job_controller.go#L472-L474

This pod doesn't have a parent, so we should just clean the finalizer.

The other case is more tricky, but I think looking at the parent to avoid incrementing the metric sounds acceptable.

Reply:
I see, but consider two jobs with custom managedBy that create pods with the built-in finalizers (which is asking for trouble anyway). 

The first job is finished, the second job is deleted by user with --cascade=orphan. 

Now, both sets of orphan pods originated from a job with custom managedBy. However, in one case we clean, the finalizers (because we no longer have parent), but not in the other (because parent still exists and has custom managedBy). So far, we didn't make that distinction.

Reply:
I just think that by making the distinctions we need to start analyzing this case by case and are likely to make a bug. This would also involve heavy testing. I think having a simple rule - pods with built-in finalizers are handled regardless of parent, as so far, makes it simple.

Reply:
> The other case is more tricky, but I think looking at the parent to avoid incrementing the metric sounds acceptable.

Probably acceptable, but I'm just concerned that making the skips for pods with built-in finalizers (which is asking for troubles anyway), if done case by case based on parent will require extensive testing, and there is almost no way to be on time by code freeze, maybe this is something to consider for beta.

However, I'm sceptical about the custom logic. I like the simplicity so far, the **built-in** controller handles pods with **built-in** finalizer, no matter of the owner. Note that currently, such pods can be created by any CRD anyway.

Reply:
> I see, but consider two jobs with custom managedBy that create pods with the built-in finalizers (which is asking for trouble anyway).

I agree that this is asking for trouble and, as a result, highly unlikely that a custom implementation would choose to use the same finalizer names.

> the built-in controller handles pods with built-in finalizer, no matter of the owner

Not the most solid argument, but good enough IMO.

I would be happy to leave this topic to be revisited in Beta.

@atiratree @soltysh can we proceed with the current implementation for alpha?

Reply:
The three of us discussed this earlier today, and we decided to skip the performance optimizations, given they don't add extra value, yet. So we'll only have the skips in `syncJob` and `syncOrphanedPods`.